crates/sui-core/src/authority_server.rs

```diff
@@ -529,15 +535,100 @@ impl ValidatorService {
+            } else {
+                None
+            }
         } else {
-            TransactionEvents::default()
+            None
         };
-        Ok(Some(HandleCertificateResponseV2 {
-            signed_effects: effects.into_inner(),
+
+        let input_objects = request
+            .include_input_objects
+            .then(|| {
+                // Note unwrapped_then_deleted contains **updated** versions.
+                let unwrapped_then_deleted_obj_ids = effects
+                    .unwrapped_then_deleted()
+                    .into_iter()
+                    .map(|k| k.0)
+                    .collect::<HashSet<_>>();
+
+                let input_object_keys = effects
+                    .input_shared_objects()
+                    .into_iter()
+                    .map(|kind| {
+                        let (id, version) = kind.id_and_version();
+                        ObjectKey(id, version)
+                    })
+                    .chain(
+                        effects
+                            .modified_at_versions()
+                            .into_iter()
+                            .map(|(object_id, version)| ObjectKey(object_id, version)),
+                    )
+                    .collect::<HashSet<_>>()
+                    .into_iter()
+                    // Unwrapped-then-deleted objects are not stored in state before the tx, so we have nothing to fetch.
+                    .filter(|key| !unwrapped_then_deleted_obj_ids.contains(&key.0))
+                    .collect::<Vec<_>>();
+
+                let input_objects = self
+                    .state
+                    .get_object_store()
+                    .multi_get_objects_by_key(&input_object_keys)?
+                    .into_iter()
+                    .enumerate()
+                    .map(|(idx, maybe_object)| {
+                        maybe_object.ok_or_else(|| {
+                            anyhow::anyhow!(
+                                "missing input object key {:?} from tx {}",
+                                input_object_keys[idx],
+                                effects.transaction_digest()
+                            )
+                        })
+                    })
+                    .collect::<anyhow::Result<Vec<_>>>()?;
+                Ok(input_objects)
+            })
+            .and_then(|maybe: anyhow::Result<_>| maybe.ok());
+
+        let output_objects = request
+            .include_output_objects
+            .then(|| {
+                let output_object_keys = effects
+                    .all_changed_objects()
+                    .into_iter()
+                    .map(|(object_ref, _owner, _kind)| ObjectKey::from(object_ref))
+                    .collect::<Vec<_>>();
+
+                let output_objects = self
+                    .state
+                    .get_object_store()
+                    .multi_get_objects_by_key(&output_object_keys)?
```

Comment:
you could use https://github.com/MystenLabs/sui/blob/f9861dd70610ee06cdb7837799b62df61a56964e/crates/sui-core/src/execution_cache.rs#L125 here if you want

Reply:
Is this case not an issue for this usecase?

```
If the object we are asking is currently locked at a future version, we know this
transaction is out-of-date and we return a ObjectVersionUnavailableForConsumption
```